File: 224.basic-calculator.cpp
/*
 * @lc app=leetcode id=224 lang=cpp
 *
 * [224] Basic Calculator
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
bool isNumeric(const string &s)
{
    if (s.size() == 0)
        return false;

    for (const char &c : s)
    {
        if (!isdigit(c))
            return false;
    }

    return true;
}

class Operand
{
public:
    virtual int eval() const { return 0; };
};

class NumericOperand : public Operand
{
private:
    int value;

public:
    NumericOperand(int v) : value(v){};
    int eval() const final;
};

int NumericOperand::eval() const { return value; }

class NestedOperand : public Operand
{
private:
    vector<Operand> operands;
    vector<char> operators;

public:
    NestedOperand() {}
    NestedOperand(Operand op) : operands({op}) {}
    int eval() const final;
    void add_operand(Operand op)
    {
        operands.push_back(op);
    }

    void add_operator(char c)
    {
        operators.push_back(c);
    }
};

int NestedOperand::eval() const
{
    if (size(operands) == size(operators))
    {
        int output = 0;

        for (int i = 0; i < size(operands); ++i)
        {
            if (operators[i] == '+')
            {
                output += operands[i].eval();
            }
            else
            {
                output -= operands[i].eval();
            }
        }

        return output;
    }

    if (size(operands) - 1 == size(operators))
    {
        int output = operands[0].eval();

        for (int i = 1; i < size(operands); ++i)
        {
            if (operators[i - 1] == '+')
            {
                output += operands[i].eval();
            }
            else
            {
                output -= operands[i].eval();
            }
        }

        return output;
    }

    return 0;
}

class Solution
{
public:
    int calculate(string s)
    {
        stack<NestedOperand> stack;
        NestedOperand base = NestedOperand();
        stack.push(base);

        for (int i = 0; i < s.size(); ++i)
        {
            string c = {s[i]};
            auto &top = stack.top();

            if (isNumeric(c))
            {
                while (i + 1 < s.size() && isdigit(s[i + 1]))
                {
                    c += s[i + 1];
                    ++i;
                }
                top.add_operand(NumericOperand(atoi(c.c_str())));
            }
            else if (c == "+" || c == "-")
            {
                top.add_operator(c[0]);
            }
            else if (c == "(")
            {
                stack.push(NestedOperand());
            }
            else if (c == ")")
            {
                int val = top.eval();
                stack.pop();
                auto &new_top = stack.top();
                new_top.add_operand(NumericOperand(val));
            }
        }

        return stack.top().eval();
    }
};
// @lc code=end

int main()
{
    Solution s;
    cout << s.calculate("1 + 1") << endl;
    return 0;
}

File: 224.basic-calculator.py
#
# @lc app=leetcode id=224 lang=python3
#
# [224] Basic Calculator
#

# @lc code=start
class Operand:
    def eval(self) -> int:
        return 0

class NumericOperand(Operand):
    def __init__(self, val: int):
        self._val = val
    
    def eval(self) -> int:
        return self._val

class NestedOperand(Operand):
    def __init__(self):
        self._operands = []
        self._operators = []
    
    def eval(self) -> int:
        if len(self._operands) - 1 == len(self._operators):
            output = self._operands[0].eval()
            
            for i in range(1, len(self._operands)):
                if self._operators[i - 1] == "+":
                    output += self._operands[i].eval()
                else:
                    output -= self._operands[i].eval()
            
            return output
        elif len(self._operands) == len(self._operators):
            output = 0
            
            for i in range(len(self._operands)):
                if self._operators[i] == "+":
                    output += self._operands[i].eval()
                else:
                    output -= self._operands[i].eval()
            
            return output
        
        return 0
    
    def add_operand(self, op: Operand):
        self._operands.append(op)
    
    def add_operator(self, op: str):
        self._operators.append(op)
    
    

class Solution:
    def calculate(self, s: str) -> int:
        stack: list[NestedOperand] = [NestedOperand()]
        
        i = 0
        while (i < len(s)):
            c = s[i]
            top = stack[-1]
            if c.isnumeric():
                while i + 1 < len(s) and s[i + 1].isnumeric():
                    c += s[i + 1]
                    i += 1
                
                top.add_operand(NumericOperand(int(c)))
            
            elif c == "+" or c == "-":
                top.add_operator(c)
            elif c == "(":
                stack.append(NestedOperand())
            elif c == ")":
                val = top.eval()
                stack.pop(-1)
                new_top = stack[-1]
                new_top.add_operand(NumericOperand(val))
        
            i += 1
        
        return stack[-1].eval()
        
# @lc code=end
if __name__ == "__main__":
    s = Solution()
    output = s.calculate("1-(-2)")
    
    print(output)

